### PR TITLE
[`core`] Add Quantization support for `dispatch_model`

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Fixes https://github.com/huggingface/accelerate/issues/1228

Before this PR, if a user wants to call `dispatch_model` on a customly created model, the model remains unquantized and leads to some errors that are hard to understand and debug. This issue has been also observed several times on other libraries that uses `accelerate` + `bitsandbytes` such as in `PEFT` that calls `dispatch_model` under the hood as well.: https://github.com/huggingface/peft/issues/187 

Now with this PR, the script below returns the expected output: 
```python
import torch
from transformers import AutoConfig, AutoModel
from transformers.utils.bitsandbytes import replace_8bit_linear

from accelerate import init_empty_weights
from huggingface_hub import hf_hub_download

from accelerate import load_checkpoint_and_dispatch
from itertools import chain

with init_empty_weights():
    model = AutoModel.from_config(AutoConfig.from_pretrained("bigscience/bloom-560m"))


model = replace_8bit_linear(model)

# For some reason replace_8bit_linear creates parameters with requires_grad=True but it's irrelevant rn
for p in model.parameters():
    p.requires_grad = False

model_path = hf_hub_download("bigscience/bloom-560m", "pytorch_model.bin")

model = load_checkpoint_and_dispatch(
    model,
    checkpoint=model_path,
    device_map="auto",
)

print(f"{model.h[0].self_attention.query_key_value.weight.device}\n{model.h[0].self_attention.query_key_value.weight.dtype}")
```

Also, this PR is fully backward compatible with `transformers` + `bitsandbytes` integration, I ran all the bnb slow tests and they pass.

cc @sgugger @pacman100

### CODE REVIEW

File: src/accelerate/utils/modeling.py [hunk 1]
```diff
@@ -156,11 +156,27 @@ def set_module_tensor_to_device(
...
+                    new_value = new_value.to(torch.float16)
                 new_value = param_cls(new_value, requires_grad=old_value.requires_grad, **kwargs).to(device)
             else:
                 new_value = param_cls(new_value, requires_grad=old_value.requires_grad).to(device)
             module._parameters[tensor_name] = new_value
 
+            if (
+                module.__class__.__name__ == "Linear8bitLt"
+                and getattr(module.weight, "SCB", None) is None
+                and module.bias is not None
```
Comment (BlackSamorez) [inline src/accelerate/utils/modeling.py:170]: Why require bias?

Comment (younesbelkada) [reply src/accelerate/utils/modeling.py:170]: Good point, this seems to be not needed indeed

Comment (BlackSamorez) [reply src/accelerate/utils/modeling.py:170]:
```suggestion
```

Comment (younesbelkada) [reply src/accelerate/utils/modeling.py:170]: I think you need more checks, I will push something now

Comment (younesbelkada) [reply src/accelerate/utils/modeling.py]: f7e9fd3

Comment (younesbelkada) [reply src/accelerate/utils/modeling.py]: Btw you can test it right away on your Kaggle notebook with:
```bash
pip install git+https://github.com/younesbelkada/accelerate.git@fix-dispatch-int9
```

File: src/accelerate/utils/modeling.py [hunk 2]
```diff
@@ -156,11 +156,24 @@ def set_module_tensor_to_device(
...
+                if new_value.dtype == torch.float32:
+                    new_value = new_value.to(torch.float16)
                 new_value = param_cls(new_value, requires_grad=old_value.requires_grad, **kwargs).to(device)
             else:
                 new_value = param_cls(new_value, requires_grad=old_value.requires_grad).to(device)
             module._parameters[tensor_name] = new_value
 
+            if module.__class__.__name__ == "Linear8bitLt" and getattr(module.weight, "SCB", None) is None:
+                # quantize only if necessary
+                if not getattr(module.weight, "SCB", None) and isinstance(device, int):
```
Comment (BlackSamorez) [inline src/accelerate/utils/modeling.py:169]: It doesn't work when device is `"cuda:0"` of `torch.device.cuda`.

Comment (younesbelkada) [reply src/accelerate/utils/modeling.py:169]: I think in this method `device` is always an int for GPUs, can you share a reproducible script?

Comment (BlackSamorez) [reply src/accelerate/utils/modeling.py:169]: It look like it isn't, when using a custom `device_map`. I'll try and create a smaller script because right now I'm working with 13B model with custom wrappers.
But from what I've seen it work if and only if devices in a custom `device_map` are `int`s

File: src/accelerate/utils/modeling.py [hunk 3]
```diff
@@ -156,11 +156,24 @@ def set_module_tensor_to_device(
...
+                    if module.bias is not None and module.bias.device.type != "meta":
+                        # if a bias exists, we need to wait until the bias is set on the correct device
+                        module = module.cuda(device)
```
Comment (BlackSamorez) [inline src/accelerate/utils/modeling.py:172]: This will fail if weight requires grad with `data set to a tensor that requires gradients must be floating point or complex dtype`. Maybe check for module.weight.requires_grad or throw own error.

File: src/accelerate/utils/modeling.py (hunk 2 above)
Comment (sgugger) [reply src/accelerate/utils/modeling.py:169]: I think you want to test device.type here (after converting the device received to a torch.device)

File: src/accelerate/utils/modeling.py (hunk 3 above)
Comment (sgugger) [inline src/accelerate/utils/modeling.py:172]: Let's use module.to(...)

File: src/accelerate/utils/modeling.py [hunk 4]
```diff
@@ -156,11 +156,24 @@ def set_module_tensor_to_device(
...
+                    elif module.bias is None:
+                        # if no bias exists, we can quantize right away
+                        module = module.cuda(device)
```
Comment (sgugger) [inline src/accelerate/utils/modeling.py:175]: Here too

**Review by sgugger — COMMENTED** [review]
Thanks for working on this!

File: src/accelerate/utils/modeling.py [hunk 5]
```diff
@@ -156,11 +156,24 @@ def set_module_tensor_to_device(
...
+
```
Comment (BlackSamorez) [inline src/accelerate/utils/modeling.py:176]: This both checks `device.type` and gets correct index from any device representation
```suggestion
            if module.__class__.__name__ == "Linear8bitLt" and getattr(module.weight, "SCB", None) is None:
                # quantize only if necessary
                device_index = torch.device(device).index if torch.device(device).type == "cuda" else None
                if not getattr(module.weight, "SCB", None) and device_index is not None:
                    if module.bias is not None and module.bias.device.type != "meta":
                        # if a bias exists, we need to wait until the bias is set on the correct device
                        module = module.cuda(device_index)
                    elif module.bias is None:
                        # if no bias exists, we can quantize right away
                        module = module.cuda(device_index)
```

File: src/accelerate/utils/modeling.py (hunk 3 above)
Comment (younesbelkada) [reply src/accelerate/utils/modeling.py]: Yes, since we place the model weights on gpu already, we need to "force quantize" the model by calling `cuda()`

File: src/accelerate/utils/modeling.py (hunk 2 above)
Comment (younesbelkada) [reply src/accelerate/utils/modeling.py]: I believe this has been addressed in [bfd7e1a](https://github.com/huggingface/accelerate/pull/1237/commits/bfd7e1afea3cee100dcbed91ae6ffd89dee09432) (also added some tests later on)

File: src/accelerate/utils/modeling.py (hunk 3 above)
Comment (younesbelkada) [reply src/accelerate/utils/modeling.py]: Let's force `module.requires_grad` to be `False` in `replace_8bit_linear`, I will make a follow up PR on `transformers` for this

Comment (younesbelkada) [reply src/accelerate/utils/modeling.py]: https://github.com/huggingface/transformers/pull/22396

**Review by sgugger — APPROVED** [review]
Looks good on my side. @BlackSamorez can you confirm this solves all your issues?

**Comment by BlackSamorez** [comment]
@sgugger yes, everything works great on my end.
Thanks everyone!
